src/assets/TokenListController.ts

```diff
@@ -0,0 +1,270 @@
+   *
+   * @param options - Constructor options
+   * @param options.interval - The polling interval, in milliseconds
+   * @param options.messenger - A reference to the messaging system
+   * @param options.state - Initial state to set on this controller
+   */
+  constructor({
+    chainId,
+    onNetworkStateChange,
+    interval = DEFAULT_INTERVAL,
+    cacheRefreshThreshold = DEFAULT_THRESHOLD,
+    messenger,
+    state,
+  }: {
+    chainId: string;
+    onNetworkStateChange: (
+      listener: (networkState: NetworkState) => void,
+    ) => void;
+    interval?: number;
+    cacheRefreshThreshold?: number;
+    messenger: RestrictedControllerMessenger<
+      typeof name,
+      GetTokenListState,
+      TokenListStateChange,
+      never,
+      never
+    >;
+    state?: Partial<TokenListState>;
+  }) {
+    super({
+      name,
+      metadata,
+      messenger,
+      state: { ...defaultState, ...state },
+    });
+    this.intervalDelay = interval;
+    this.cacheRefreshThreshold = cacheRefreshThreshold;
+    this.chainId = chainId;
+    onNetworkStateChange((networkState) => {
+      this.chainId = networkState.provider.chainId;
+    });
+  }
+
+  /**
+   * Start polling for the token list
+   */
+  async start() {
+    await this.startPolling();
+  }
+
+  /**
+   * Stop polling for the token list
+   */
+  stop() {
+    this.stopPolling();
+  }
+
+  /**
+   * Prepare to discard this controller.
+   *
+   * This stops any active polling.
+   */
+  destroy() {
+    super.destroy();
+    this.stopPolling();
+  }
+
+  private stopPolling() {
+    if (this.intervalId) {
+      clearInterval(this.intervalId);
+    }
+  }
+
+  /**
+   * Starts a new polling interval
+   */
+  private async startPolling(): Promise<void> {
+    await safelyExecute(() => this.fetchTokenList());
+    this.intervalId = setInterval(async () => {
+      await safelyExecute(() => this.fetchTokenList());
+    }, this.intervalDelay);
+  }
+
+  /**
+   * Fetching token list from the Token Service API
+   */
+  async fetchTokenList(): Promise<void> {
+    const releaseLock = await this.mutex.acquire();
+    try {
+      const tokensFromAPI: Token[] = await safelyExecute(() =>
+        this.fetchFromCache(),
+      );
+      const { tokensChainsCache } = this.state;
+      const tokenList: TokenMap = {};
+
+      // filtering out tokens with less than 2 occurences
+      const filteredTokenList = tokensFromAPI.filter(
+        (token) => token.occurrences >= 2,
+      );
+      // removing the tokens with symbol conflicts
+      const symbolsList = filteredTokenList.map((token) => token.symbol);
+      const duplicateSymbols = [
+        ...new Set(
+          symbolsList.filter(
+            (symbol, index) => symbolsList.indexOf(symbol) !== index,
+          ),
+        ),
+      ];
+      const uniqueTokenList = filteredTokenList.filter(
+        (token) => !duplicateSymbols.includes(token.name),
+      );
+      for (const token of uniqueTokenList) {
```

Comment:
Is this arbitrarily privileging tokens with duplicated symbols that have lower indices in what's returned from our API?

Reply:
For the MVP we are not showing the tokens with conflicting symbols, we are removing all tokens with duplicate symbols.